builtin/v10/migration/miner.go

```diff
@@ -0,0 +1,437 @@
+	adt9 "github.com/filecoin-project/go-state-types/builtin/v9/util/adt"
+)
+
+// The minerMigrator performs the following migrations:
+// FIP-0047 Adds ProofExpiration to SectorOnChainInfo, and FaultySectors to the ExpirationSet of each partition
+
+type minerMigrator struct {
+	emptyDeadlineV9   cid.Cid
+	emptyDeadlinesV9  cid.Cid
+	emptyDeadlineV10  cid.Cid
+	emptyDeadlinesV10 cid.Cid
+	emptyPartitions   cid.Cid
+	OutCodeCID        cid.Cid
+}
+
+func newMinerMigrator(ctx context.Context, store cbor.IpldStore, outCode cid.Cid) (*minerMigrator, error) {
+	ctxStore := adt9.WrapStore(ctx, store)
+
+	edv9, err := miner9.ConstructDeadline(ctxStore)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to construct empty deadline v8: %w", err)
+	}
+
+	edv9cid, err := store.Put(ctx, edv9)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to put empty deadline v8: %w", err)
+	}
+
+	edsv9 := miner9.ConstructDeadlines(edv9cid)
+	edsv9cid, err := store.Put(ctx, edsv9)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to construct empty deadlines v8: %w", err)
+	}
+
+	edv10, err := miner10.ConstructDeadline(ctxStore)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to construct empty deadline v9: %w", err)
+	}
+
+	edv10cid, err := store.Put(ctx, edv10)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to put empty deadline v9: %w", err)
+	}
+
+	edsv10 := miner10.ConstructDeadlines(edv10cid)
+	edsv10cid, err := store.Put(ctx, edsv10)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to construct empty deadlines v9: %w", err)
+
+	}
+
+	emptyPartitionsArrayCid, err := adt10.StoreEmptyArray(ctxStore, miner10.DeadlinePartitionsAmtBitwidth)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to construct empty partitions array: %w", err)
+	}
+
+	return &minerMigrator{
+		emptyDeadlineV9:   edv9cid,
+		emptyDeadlinesV9:  edsv9cid,
+		emptyDeadlineV10:  edv10cid,
+		emptyDeadlinesV10: edsv10cid,
+		emptyPartitions:   emptyPartitionsArrayCid,
+		OutCodeCID:        outCode,
+	}, nil
+}
+
+func (m minerMigrator) MigratedCodeCID() cid.Cid {
+	return m.OutCodeCID
+}
+
+func (m minerMigrator) MigrateState(ctx context.Context, store cbor.IpldStore, in migration.ActorMigrationInput) (*migration.ActorMigrationResult, error) {
+	var inState miner9.State
+	if err := store.Get(ctx, in.Head, &inState); err != nil {
+		return nil, xerrors.Errorf("getting miner state: %w", err)
+	}
+	var inInfo miner9.MinerInfo
+	if err := store.Get(ctx, inState.Info, &inInfo); err != nil {
+		return nil, xerrors.Errorf("getting miner info: %w", err)
+	}
+	wrappedStore := adt9.WrapStore(ctx, store)
+
+	newSectors, err := migrateSectorsWithCache(ctx, wrappedStore, in.Cache, in.Address, inState.Sectors)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to migrate sectors for miner: %s: %w", in.Address, err)
+	}
+
+	newDeadlines, err := m.migrateDeadlines(ctx, wrappedStore, in.Cache, inState.Deadlines)
```

Comment:
We have `WithCache` for sectors because we have distinct behaviour with and without the cache. I think I'll keep this naming scheme for sectors and partitions because they only have one behaviour?